command/agent/cache/cacheboltdb/bolt.go

```diff
@@ -0,0 +1,333 @@
+)
+
+// BoltStorage is a persistent cache using a bolt db. Items are organized with
+// the encryption key as the top-level bucket, and then leases and tokens are
+// stored in sub buckets.
+type BoltStorage struct {
+	db         *bolt.DB
+	rootBucket string
+	logger     hclog.Logger
+	keyMgr     crypto.KeyManager
+	aad        string
+}
+
+// BoltStorageConfig is the collection of input parameters for setting up bolt
+// storage
+type BoltStorageConfig struct {
+	Path       string
+	Logger     hclog.Logger
+	AAD        string
+	KeyManager crypto.KeyManager
+}
+
+// NewBoltStorage opens a new bolt db at the specified file path and returns it.
+// If the db already exists the buckets will just be created if they don't
+// exist.
+func NewBoltStorage(config *BoltStorageConfig) (*BoltStorage, error) {
+	dbPath := filepath.Join(config.Path, DatabaseFileName)
+	db, err := bolt.Open(dbPath, 0600, &bolt.Options{Timeout: 1 * time.Second})
+	if err != nil {
+		return nil, err
+	}
+	err = db.Update(func(tx *bolt.Tx) error {
+		return createBoltSchema(tx)
+	})
+	if err != nil {
+		return nil, err
+	}
+	bs := &BoltStorage{
+		db:         db,
+		rootBucket: rootBucketName,
+		logger:     config.Logger,
+		keyMgr:     config.KeyManager,
+		aad:        config.AAD,
+	}
+	if config.KeyManager != nil {
+		bs.SetKeyMgr(config.KeyManager)
+	}
+	return bs, nil
+}
+
+func createBoltSchema(tx *bolt.Tx) error {
+	_, err := tx.CreateBucketIfNotExists([]byte(KeyBucket))
+	if err != nil {
+		return fmt.Errorf("failed to create key bucket: %w", err)
+	}
+	root, err := tx.CreateBucketIfNotExists([]byte(rootBucketName))
+	if err != nil {
+		return fmt.Errorf("failed to create bucket %s: %w", rootBucketName, err)
+	}
+	_, err = root.CreateBucketIfNotExists([]byte(TokenType))
+	if err != nil {
+		return fmt.Errorf("failed to create token sub-bucket: %w", err)
+	}
+	_, err = root.CreateBucketIfNotExists([]byte(AuthLeaseType))
+	if err != nil {
+		return fmt.Errorf("failed to create auth lease sub-bucket: %w", err)
+	}
+	_, err = root.CreateBucketIfNotExists([]byte(SecretLeaseType))
+	if err != nil {
+		return fmt.Errorf("failed to create secret lease sub-bucket: %w", err)
+	}
+
+	// check and set file version in the root bucket
+	version := root.Get([]byte(storageVersionKey))
+	switch {
+	case version == nil:
+		err = root.Put([]byte(storageVersionKey), []byte(storageVersion))
+		if err != nil {
+			return fmt.Errorf("failed to set storage version: %w", err)
+		}
+	case string(version) != storageVersion:
+		return fmt.Errorf("storage migration from %s to %s not implemented", string(version), storageVersion)
+	}
+	return nil
+}
+
+// SetKeyMgr sets the encryption for a bolt storage
+func (b *BoltStorage) SetKeyMgr(km crypto.KeyManager) {
+	b.keyMgr = km
+}
+
+// Set an index in bolt storage
+func (b *BoltStorage) Set(ctx context.Context, id string, plainText []byte, indexType string) error {
+	cipherText, err := b.keyMgr.Encrypt(ctx, plainText, []byte(b.aad))
```

Comment:
This feels like key operation is leaking a bit into the `KeyManager` interface. `KeyManager` should just be in charge of the key management such as retrieval, renewal/rewrap, etc.`BoltStorage` would have `wrapping.Wrapper` as a field to handle the encryption. Instead of passing in KeyManager, you pass in the key material and aad.

Upon reading the code a bit further, I feel like the correct layout would be to have an implementation of KeyManager on BoltDB. The `KubeEncryptionKey` implementation has nothing that's specific to k8s. On the agent config side, we can still call it `crypto "kubernetes"`, but this is really being able to get and set the key from and to BoltDB (and in fact we're already exposing methods such as `GetKey` and `SetKey` which is pretty much what KeyManager should be doing).

Reply:
Originally the KubeEncryptionKey code did have Kube specific code, namely getting the service account JWT which we use for AAD. We can add that back in so the caller doesn't need to know about it and that would make it more Kube specific.